### ISSUE TITLE
[Randgen] TIDB panic running optimizer_subquery.yy test

### ISSUE THREAD

**Issue by sokoow** [pingcap/tidb#4996]
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
- download randgen https://launchpad.net/randgen
- build TIDB 1.0.1 (1.1.0 doesn't build sucessfully)
- run desired test: ```./gentest.pl --dsn dbi:mysql:host=127.0.0.1:port=4000:user=root:database=test --grammar=conf/optimizer_subquery.yy --gendata --thread=1```

2. What did you expect to see?
Test passing

3. What did you see instead?
TIDB panic with :

```
panic: runtime error: index out of range

goroutine 99985 [running]:
github.com/pingcap/tidb/expression.(*Column).EvalInt(0xc4279d7110, 0xc4214b9d40, 0x7, 0x7, 0xc42833ef60, 0x0, 0x18bea00, 0x0, 0x0)
        /go/src/github.com/pingcap/tidb/expression/column.go:202 +0x165
github.com/pingcap/tidb/expression.(*builtinLogicOrSig).evalInt(0xc426de8b40, 0xc4214b9d40, 0x7, 0x7, 0x893f24, 0xc426de8b40, 0x18ac880, 0xc423240500)
        /go/src/github.com/pingcap/tidb/expression/builtin_op.go:123 +0x111
github.com/pingcap/tidb/expression.(*ScalarFunction).EvalInt(0xc426de8b80, 0xc4214b9d40, 0x7, 0x7, 0xc42833ef60, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/pingcap/tidb/expression/scalar_function.go:223 +0x52
github.com/pingcap/tidb/expression.(*ScalarFunction).Eval(0xc426de8b80, 0xc4214b9d40, 0x7, 0x7, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/pingcap/tidb/expression/scalar_function.go:193 +0x305
github.com/pingcap/tidb/expression.EvalBool(0xc421845920, 0x1, 0x1, 0xc4214b9d40, 0x7, 0x7, 0x18ac880, 0xc423240500, 0x0, 0xc421d66800, ...)
        /go/src/github.com/pingcap/tidb/expression/expression.go:103 +0xac
github.com/pingcap/tidb/executor.(*HashJoinExec).constructMatchedRows(0xc42181e180, 0xc424d643c0, 0xc426cba480, 0x5, 0x5, 0x5, 0x18ac880, 0xc423240500, 0xc420b77701, 0x0)
        /go/src/github.com/pingcap/tidb/executor/join.go:388 +0x3fb
github.com/pingcap/tidb/executor.(*HashJoinExec).joinOneBigRow(0xc42181e180, 0xc424d643c0, 0xc426cba480, 0x5, 0x5, 0xc4238770b0, 0xc420b776c8)
        /go/src/github.com/pingcap/tidb/executor/join.go:347 +0x3e1
github.com/pingcap/tidb/executor.(*HashJoinExec).runJoinWorker(0xc42181e180, 0x1)
        /go/src/github.com/pingcap/tidb/executor/join.go:316 +0x2a9
created by github.com/pingcap/tidb/executor.(*HashJoinExec).prepare
        /go/src/github.com/pingcap/tidb/executor/join.go:250 +0x52a
panic: runtime error: index out of range

goroutine 99904 [running]:
github.com/pingcap/tidb/expression.(*Column).EvalInt(0xc4279d6ea0, 0xc4203bab60, 0x7, 0x7, 0xc42833ef60, 0x0, 0x18bea00, 0x0, 0x0)
        /go/src/github.com/pingcap/tidb/expression/column.go:202 +0x165
github.com/pingcap/tidb/expression.(*builtinLogicOrSig).evalInt(0xc426de8a40, 0xc4203bab60, 0x7, 0x7, 0x893f24, 0xc426de8a40, 0x18ac880, 0xc423240500)
        /go/src/github.com/pingcap/tidb/expression/builtin_op.go:123 +0x111
github.com/pingcap/tidb/expression.(*ScalarFunction).EvalInt(0xc426de8a80, 0xc4203bab60, 0x7, 0x7, 0xc42833ef60, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/pingcap/tidb/expression/scalar_function.go:223 +0x52
github.com/pingcap/tidb/expression.(*ScalarFunction).Eval(0xc426de8a80, 0xc4203bab60, 0x7, 0x7, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/pingcap/tidb/expression/scalar_function.go:193 +0x305
github.com/pingcap/tidb/expression.EvalBool(0xc421845910, 0x1, 0x1, 0xc4203bab60, 0x7, 0x7, 0x18ac880, 0xc423240500, 0x0, 0x0, ...)
        /go/src/github.com/pingcap/tidb/expression/expression.go:103 +0xac
github.com/pingcap/tidb/executor.(*HashJoinExec).constructMatchedRows(0xc42181e180, 0xc424d642a0, 0xc426cba360, 0x5, 0x5, 0x5, 0x18ac880, 0xc423240500, 0xc425914701, 0x0)
        /go/src/github.com/pingcap/tidb/executor/join.go:388 +0x3fb
github.com/pingcap/tidb/executor.(*HashJoinExec).joinOneBigRow(0xc42181e180, 0xc424d642a0, 0xc426cba360, 0x5, 0x5, 0xc423877080, 0xc425319700)
        /go/src/github.com/pingcap/tidb/executor/join.go:347 +0x3e1
github.com/pingcap/tidb/executor.(*HashJoinExec).runJoinWorker(0xc42181e180, 0x0)
        /go/src/github.com/pingcap/tidb/executor/join.go:316 +0x2a9
created by github.com/pingcap/tidb/executor.(*HashJoinExec).prepare
        /go/src/github.com/pingcap/tidb/executor/join.go:250 +0x52a
```

4. What version of TiDB are you using (`tidb-server -V`)?
1.0.1

**Comment by shenli** [comment]
@sokoow Thanks for your feedback!

**Comment by XuHuaiyu** [comment]
same issue as #4997

**Comment by XuHuaiyu** [comment]
fixed in #5086
